rewrite-java/src/main/java/org/openrewrite/java/cleanup/RemoveUnusedAssignments.java

```diff
@@ -0,0 +1,160 @@
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.cleanup;
+
+import org.openrewrite.*;
+import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.java.DeleteStatement;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.tree.Expression;
+import org.openrewrite.java.tree.J;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Logic to determine if an assignment to an identifier has any read operations performed on it.
+ * Any read operations performed on an assignment LHS (left-hand side) means the assignment RHS (right-hand side) is used.
+ * <p>
+ * If an assignment is overwritten by a reassignment later, the previous assignment may need to be removed.
+ * In that case, we need to look at whether there are read operations performed on the LHS assignment
+ * in the scope between the assignment and the reassignment.
+ * <p>
+ * If there are read operations, we should not remove the assignment. But if there are not any LHS read operations
+ * between the first assignment and the reassignment, then we can prune the original assignment. It's dead code (gross!).
+ */
+@Incubating(since = "7.10.0")
+@SuppressWarnings("AnonymousInnerClassMayBeStatic")
+public class RemoveUnusedAssignments extends Recipe {
+    @Override
+    public String getDisplayName() {
+        return "Remove unused assignments";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Calculating a value followed by overwriting it could indicate a coding error, or at least a waste of resources.";
+    }
+
+    @Override
+    public Set<String> getTags() {
+        return Collections.singleton("RSPEC-1854");
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+            private final Map<String, List<J.Assignment>> identifierAssignments = new HashMap<>();
+            private final JavaIsoVisitor<ExecutionContext> collectAssignments = new JavaIsoVisitor<ExecutionContext>() {
+                @Override
+                public J.Assignment visitAssignment(J.Assignment assignment, ExecutionContext ctx) {
+                    if (assignment.getVariable() instanceof J.Identifier) {
+                        identifierAssignments.computeIfAbsent(((J.Identifier) assignment.getVariable()).getSimpleName(), f -> new ArrayList<>()).add(assignment);
+                    }
+                    return super.visitAssignment(assignment, ctx);
+                }
+            };
```

Comment:
This sure looks like it would be broken for sources where different lexical scopes used identifiers with coincidentally matching names.

void foo() {
    int a = 1;
}

void bar() {
    int a = 1;
}

Whether or not I'm right that there's a problem here, there should be tests covering this very common scenario and similar.